ibis/pandas/execution/tests/test_window.py

```diff
@@ -582,3 +584,124 @@ def count_both(v1, v2):
         t[control].count().over(row_window).execute(),
         check_names=False,
     )
+
+
+def test_custom_window_udf(t, df):
+    """ Test implementing  a (dummy) custom window.
+
+    This test covers the advance use case to support custom window with udfs.
+
+    Note that method used in this example (e.g, get_preceding, get_aggcontext)
+    are unstable developer API, not stable public API.
+    """
+
+    @reduction(input_type=[dt.float64], output_type=dt.float64)
+    def my_sum(v):
+        return v.sum()
+
+    class CustomInterval(object):
+        def __init__(self, value):
+            self.value = value
+
+        # These are necessary because ibis.expr.window
+        # will compare preceding and following
+        # with 0 to see if they are valid
+        def __lt__(self, other):
+            return self.value < other
+
+        def __gt__(self, other):
+            return self.value > other
+
+    class CustomWindow(ibis.expr.window.Window):
+        """ This is a dummy custom window that return n preceding rows
+        where n is defined by CustomInterval.value."""
+
+        def _replace(self, **kwds):
+            new_kwds = dict(
+                group_by=kwds.get('group_by', self._group_by),
+                order_by=kwds.get('order_by', self._order_by),
+                preceding=kwds.get('preceding', self.preceding),
+                following=kwds.get('following', self.following),
+                max_lookback=kwds.get('max_lookback', self.max_lookback),
+                how=kwds.get('how', self.how),
+            )
+            return CustomWindow(**new_kwds)
+
+    class CustomAggContext(AggregationContext):
+        def __init__(
+            self, parent, group_by, order_by, dtype, max_lookback, preceding
+        ):
+            super().__init__(
+                parent=parent,
+                group_by=group_by,
+                order_by=order_by,
+                dtype=dtype,
+                max_lookback=max_lookback,
+            )
+            self.preceding = preceding
+
+        def agg(self, grouped_data, function, *args, **kwargs):
+            upper_indices = pd.Series(range(len(self.parent) + 1)) + 1
```

Comment:
May be a bit more clear:
```suggestion
            upper_indices = pd.Series(range(1, len(self.parent) + 2))
```